types/connect/index.d.ts

```diff
@@ -1,94 +1,163 @@
 
 /**
- * Create a new connect server.
- * @public
+ * When the `connect` module is required, a function is returned that will construct a new app when called.
+ * This app will store all the middleware added and is, itself, a function.
+ *
+ * The core of Connect is "using" middleware. Middleware are added as a "stack" where incoming
+ * requests will execute each middleware one-by-one until a middleware does not call `next()` within it.
+ *
+ * There are special cases of "error-handling" middleware. These are middleware where the
+ * function takes exactly 4 arguments. When a middleware passes an error to `next`, the app will
+ * proceed to look for the error middleware that was declared after that middleware and invoke it,
+ * skipping any error middleware above that middleware and any non-error middleware below.
+ *
+ * The last step is to actually use the Connect app in a server. The `.listen()` method is a
+ * convenience to start a HTTP server (and is identical to the `http.Server`'s `listen` method).
+ * The app itself is really just a function with three arguments, so it can also be handed to
+ * `.createServer()` in Node.js.
  */
-declare function createServer(): createServer.Server;
+declare function createServer(): connect.Server;
 
 declare namespace createServer {
+    export interface Request extends http.IncomingMessage {
+        /**
+         * The request URL, altered to remove the route part. For example, if the handle function is used
+         * at the route `/foo`, the request for `/foo/bar` will invoke it with `req.url === '/bar'` and
+         * `req.originalUrl === '/foo/bar'`.
+         */
+        url: string;
+        /**
+         * The original URL of the request, unaltered.
+         */
+        originalUrl: string;
+    }
+
     export type ServerHandle = HandleFunction | http.Server;
 
-    type NextFunction = (err?: any) => void;
+    type NextFunction = (
+        /** An optional error, to make the app skip to error-handling middleware. */ err?: any
+    ) => void;
 
-    export type SimpleHandleFunction = (req: http.IncomingMessage, res: http.ServerResponse) => void;
-    export type NextHandleFunction = (req: http.IncomingMessage, res: http.ServerResponse, next: NextFunction) => void;
-    export type ErrorHandleFunction = (err: any, req: http.IncomingMessage, res: http.ServerResponse, next: NextFunction) => void;
-    export type HandleFunction = SimpleHandleFunction | NextHandleFunction | ErrorHandleFunction;
+    /** @deprecated `next` is always available to the handler */ export type SimpleHandleFunction = NextHandleFunction;
+    export type NextHandleFunction =
+        (req: Request | http.IncomingMessage, res: http.ServerResponse, next: NextFunction) => void;
+    export type ErrorHandleFunction =
+        (err: any, req: Request | http.IncomingMessage, res: http.ServerResponse, next: NextFunction) => void;
+    export type HandleFunction = NextHandleFunction | ErrorHandleFunction;
```

Comment:
Param `req` was originally changed to be `connect.Request` (with originalUrl) but I had to keep `http.IncomingMessage` too to avoid breaking a test in another package which referenced the `HandleFunction` type directly by name (even though its a subtype).
I should probably have `@deprecated` all of those `*HandleFunction` too anyways.